repository/src/main/java/org/alfresco/repo/security/authentication/identityservice/SpringBasedIdentityServiceFacade.java

```diff
@@ -1,344 +1,352 @@
+
+    @Override
+    public AccessTokenAuthorization authorize(AuthorizationGrant authorizationGrant)
+    {
+        final AbstractOAuth2AuthorizationGrantRequest request = createRequest(authorizationGrant);
+        final OAuth2AccessTokenResponseClient client = getClient(request);
+
+        final OAuth2AccessTokenResponse response;
+        try
+        {
+            response = client.getTokenResponse(request);
+        }
+        catch (OAuth2AuthorizationException e)
+        {
+            LOGGER.debug("Failed to authorize against Authorization Server. Reason: " + e.getError() + ".");
+            throw new AuthorizationException("Failed to obtain access token. " + e.getError(), e);
+        }
+        catch (RuntimeException e)
+        {
+            LOGGER.warn("Failed to authorize against Authorization Server. Reason: " + e.getMessage());
+            throw new AuthorizationException("Failed to obtain access token.", e);
+        }
+
+        return new SpringAccessTokenAuthorization(response);
+    }
+
+    @Override
+    public Optional<OIDCUserInfo> getUserInfo(String tokenParameter, String principalAttribute)
+    {
+        return Optional.ofNullable(tokenParameter)
+                .filter(Predicate.not(String::isEmpty))
+                .flatMap(token -> Optional.ofNullable(clientRegistration)
+                        .map(ClientRegistration::getProviderDetails)
+                        .map(ClientRegistration.ProviderDetails::getUserInfoEndpoint)
+                        .map(ClientRegistration.ProviderDetails.UserInfoEndpoint::getUri)
+                        .flatMap(uri -> {
+                            try
+                            {
+                                return Optional.of(
+                                        new UserInfoRequest(new URI(uri), new BearerAccessToken(token)).toHTTPRequest().send());
+                            }
+                            catch (IOException | URISyntaxException e)
+                            {
+                                LOGGER.warn("Failed to get user information. Reason: " + e.getMessage());
+                                return Optional.empty();
+                            }
+                        })
+                        .flatMap(httpResponse -> {
+                            try
+                            {
+                                UserInfoResponse userInfoResponse = UserInfoResponse.parse(httpResponse);
+
+                                if (userInfoResponse instanceof UserInfoErrorResponse)
+                                {
+                                    UserInfoErrorResponse userInfoErrorResponse = (UserInfoErrorResponse) userInfoResponse;
+                                    String errorMessage = userInfoErrorResponse.getErrorObject().getDescription();
+                                    LOGGER.warn("User Info Request failed: " + errorMessage);
+                                    throw new UserInfoException(errorMessage);
```

Comment:
Do we need to throw the error exception? Can't we extract the error message and use the LOGGER.warn to show the error and then return the empty Optional value as done for ParseException?

Reply:
I agree with @kavitshah-gl .

Reply:
> Do we need to throw the error exception? Can't we extract the error message and use the LOGGER.warn to show the error and then return the empty Optional value as done for ParseException?

If we send the empty response then the Client SDK is not receiving the error at all which does not help users. Further if we continue with the UserInfoErrorResponse further to the map method then it gives ClassCastException. This is happening since UserInfoResponse can be converted to UserInfoSuccessResponse but UserInfoErrorResponse can't ne converted to UserInfoSuccessResponse and in the map it tries to do that . Thus rather than Spring throw the ClassCastException we are throwing UserInfoException which is of type IdentityServiceFacadeException which is caught gracefully in authenticateImpl method.

Reply:
Okay I think throwing UserInfoException is fine since we are catching it.